pkg/identity/api/v1/service.go

```diff
@@ -42,18 +42,36 @@ func (s *Service) Close() {
 }
 
 /*
-Algorithm:
-
-Start transaction
-
- 1. Insert the update into the inbox_log table
- 2. Read the log for the inbox_id, ordering by sequence_id
- 3. If the log has more than 256 entries, abort the transaction.
- 3. Validate it sequentially. If failed, abort the transaction.
+Properties we want on the inbox log:
+ 1. Updates come in and are assigned sequence numbers in some order
+ 2. Updates are not visible to API consumers until they have been
+    validated and the address_log table has been updated
+ 3. If you read once, and then read again:
+    a. The second read must have all of the updates from the first read
+    b. New updates from the second read *cannot* have a lower sequence
+    number than the latest sequence number from the first read
+    c. This only applies to reads *on a given inbox*. Ordering between
+    different inboxes does not matter.
+
+For the address log, strict ordering/strong consistency is not required
+across inboxes. Resolving eventually to any non-revoked inbox_id is
+acceptable.
+
+Algorithm for PublishIdentityUpdate:
+
+Start transaction (SERIALIZABLE isolation level)
+
+ 1. Read the log for the inbox_id, ordering by sequence_id
+    - Use FOR UPDATE to block other transactions on the same inbox_id
+    if the log is non-empty
+ 2. If the log has 256 or more entries, abort the transaction.
+ 3. Concatenate the update in-memory and validate it sequentially. If
+    failed, abort the transaction.
  4. For each affected address:
     a. Insert or update the record with (address, inbox_id) into
     the address_log table. Update the sequence_id if it is
     higher
+ 5. Insert the update into the inbox_log table
```

Comment:
I think 5 has to come before 4, since the address log contains references to the sequence id from the inbox_log.

Reply:
I don't think it matters because the transaction succeeds or fails atomically, but happy to reorder it for clarity

Reply:
I just mean that you need to know the `sequence_id` before you can do anything with the address log, since the address log has an `association_sequence_id` field

Reply:
You're totally right